src/MCPClient/lib/clientScripts/pid_declaration.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python2
+# -*- coding: utf8
+
+"""pid_declaration.py
+
+Given an identifiers.json file, supplying third-party persistent identifiers
+(PIDS), such as handle (hdl) or doi, identifiers; associate those identifiers
+with the objects in the transfer, so be translated to PREMIS objects in the
+AIP METS.
+"""
+from __future__ import unicode_literals
+from functools import wraps
+import json
+import os
+import sys
+
+import django
+
+django.setup()
+
+from main.models import Directory, File, SIP
+
+from pid_component_helpers import (
+    _add_custom_pid_to_mdl_identifiers,
+    AssignPIDsException,
+)
+from sanitize_names import sanitizeName
+
+
+class DeclarePIDsException(Exception):
+    """Exception to raise if there's a more fundamental issue with the running
+    of this script and we want to exit and stop the workflow from continuing.
+    """
+
+
+class DeclarePIDsWarning(Exception):
+    """Exception to raise when we have problems trying to open identifiers.json
+    or the file doesn't exist.
+    """
+
+    exit_code = 0
+
+
+class DeclarePIDs(object):
+    """Class to wrap PID declaration features and provide some mechanism of
+    recording state.
+    """
+
+    SIP_IDENTIFIER = "SIP"
+    SIP_DIRECTORY = "%SIPDirectory%"
+
+    def __init__(self, job):
+        """Class constructor"""
+        self.job = job
+        self.identifier_count = 0
+        self.actual_count = 0
+
+    def exit_on_known_exception(func):
+        """Decorator to allows us to raise an exception but still exit-zero when
+        the exception is cleaner to return than ad-hoc integer values.
+        """
+
+        @wraps(func)
+        def wrapped(*_args, **kwargs):
+            try:
+                func(*_args, **kwargs)
+            except (DeclarePIDsWarning) as err:
+                return err.exit_code
+
+        return wrapped
+
+    def _fixup_path_input_by_user(self, path):
+        """Fix-up paths submitted by a user, e.g. in custom structmap examples
+        so that they don't have to anticipate the Archivematica normalization
+        process.
+        """
+        return "{}{}".format(
+            self.SIP_DIRECTORY,
+            os.path.join(
+                "",
+                *[sanitizeName(name.encode("utf8")) for name in path.split(os.path.sep)]
+            ),
+        )
+
+    def _log_identifer_to_stdout(self, mdl, identifier_type, identifier):
+        """Create a consistent string to log information about our identifiers
+        to stdout for the user.
+        """
+        try:
+            if isinstance(mdl, SIP):
+                msg = "Identifier {}: {} added for: {}: {}".format(
+                    identifier_type,
+                    identifier,
+                    "SIP",
+                    os.path.basename(os.path.split(mdl.currentpath)[0]),
```

Comment:
It seems like we can use the `__unicode__` method defined on the models which already includes path information as you've done in the KeyError case? Or is there a need for this specific format?

Reply:
I adopted this approach as a matter of style perhaps, vs. substance. The output will look like this:
```
Identifier EXID: https://example.com/file/id/2026496916 added for: File: MARBLES.TGA
```
vs.

```
Identifier EXID: https://example.com/file/id/2026496916 added to File efd059dd-7050-4bf4-bae5-bf1bd47082f2:%transferDirectory%objects/artwork/MARBLES.TGA now at %SIPDirectory%objects/artwork/MARBLES.TGA
```

I thought it a bit cleaner for the user, but what do you think?

Reply:
kk, makes sense.